### DISCUSSION TITLE
VSCode workflow

### DISCUSSION THREAD

**Discussion by jvallade** [ThalesGroup/xsmp-modeler-core#43] [Ideas]
I start a new thread as a discussion, to see if we can reach a complete workflow using VSCode (no Eclipse install 😄).

I've succeeded in installing the extension using the provided vsix file. I'm also able to create a simple xsmp file. 

A first minor issue I've noticed (that should probably be a proper Github issue): each time the syntax is wrong (e.g during the definition of a new attribute / operation / etc), an error is issued by the LSP server and a pop-up appears. I'm not really familiar with the LSP protocol but it should be possible to return the fact that the syntax is wrong, without having this pop-up (that usually appear only when the extension crashed). 

Now, about the workflow. I've created a xsmp file, but I'm a bit stuck. As far as I understand, the support in VSCode stops here anyway at the moment. My first intuition is to try and use the CLI for the rest of the workflow. Sadly, I was not able to find the jar file mentioned in the documentation... Maybe it could be part of the releases somehow ? (as it seems to me that it could also be another solution beside the Eclipse plugin and the VSCode extension).

One of the suggestions I have in mind for the VSCode integration is to make use of the [VSCode commands](https://code.visualstudio.com/api/extension-guides/command) to integrate with the rest of the workflow. Commands to generate smpcat file, C++ code, unit tests could be created. Maybe even to initialize a new project with the proper project layout. 

I know I'm pushing in a direction that was not really foreseen when you started the project, but I'm convinced that having a complete VSCode support will greatly help targeting more people, especially people like me that have not used Eclipse for several years and don't really want to come back 😄. Also, if I find the time (which is far from sure), I would love to contribute a bit, but at the moment I feel like I miss too much information on the architecture of the product to know where to start. Again, if you want, I'm available to make a small call that might speed up a bit the on-boarding 😉

**Comment by ydaveluy** [comment]
For the popup it is very easy to fix.
This VSCode extension only support the editor with formatting, content assist, and validation.
The generation is not provided yet. The command mechanism seems promising to run different code generators on the fly.
The CLI are not published in the release. Each CLI is 35Mo and there are 4 CLI at the moment. I don't know if there are some limit on releases size on GitHub.
I'm not familiar with VSCode and don't know all the features of this framework.
It would be great to exchange on this topic.

Yannick

**Comment by ydaveluy** [comment]
You can find a new temporary version https://github.com/ThalesGroup/xsmp-modeler-core/releases/download/v1.1.0-rc3_tmp/xsmp-modeler-v1.1.0-rc3_tmp.vsix
I added 4 commands to run each of the generators.
The popup windows should disappear in case the syntax is wrong.

In Eclipse we have a per project configuration with a list of enabled generators. When a xsmpcat file is saved the configured generator are automatically run in background. Maybe we could have the same mechanism in VSCode instead or running each generator separately.

I have no idea on how to create a project wizard or snippet for now

**Reply by jvallade** [reply]
Thanks for the quick update ! I'll try it ASAP 😉 
As for the actions to be performed on save, I'm not sure how it could be achieved, I'll have a look, but you're right it could be yet another improvement.

**Reply by jvallade** [reply]
I could not wait, and just tested it. It works like a charm 😉 
Thanks again. 

Edit : I just noticed I'm still getting the pop-up when the syntax is not correct. Let's discuss that next week, no need to rush for a fix.

**Comment by ydaveluy** [comment]
Hello @jvallade 

We just finalized the rc6. The Eclipse installation should be fixed.

The vscode extension has been improved : 
 - wizard to create a new Xsmp project
 - Custom project settings in .xsmp/settings.json : possible to configure the profile, active tools, source folders, project dependencies and automatic generation
 - Similar behavior to Eclipse : source files are generated when xsmpcat files are saved
 - a "generate" command is provided (useful when automatic generation is disabled)

The conversion of smpcat files to xsmpcat files is not yet supported -> to be done in Eclipse.

The C++ build with xsmp-sdk profile should works with default cmake configuration

**Reply by jvallade** [reply]
Thanks! I did not have time to play with it lately... Too many things going on. I'll try to have a look at the latest version when I find some time ☺️
